Feature/tsg without rich peak

enetz opened the pull request:
An alternative TheoreticalSpectrumGenerator, that does not use RichPeaks
and instead stores String annotations and peak charges in StringDataArrays and IntegerDataArrays.

It is meant to replace the old TSG in the long run.
Right now some functions, that do not use the metainfo anyway, only accept RichPeakSpectrum or
need templates to allow RichPeakSpectrum and PeakSpectrum, because the old TSG always generates RichPeakSpectra.
Also the MetaValueInterface is slower than working with Data Arrays.

Using this alternative could simplify some functions and improve performance.

A short benchmark run: generating 1 million spectra with metainfo
Old TSG:     213.56 sec
New TSG :  176.04 sec

hroest commented:
this sounds really great, thanks for the contribution. I have some issue with the PR though: it seems you just leave the old one in, could we make a decision for the new or old TSG and then keep only one? I dont think that adding a second version of a class is the right way to go and just confuses people. Either the new way is better and replaces the old one or it is not.

enetz commented:
Yes, this one should be better and is meant to replace the old one, but there are some tools and functions that depend on the old one and expect RichPeakSpectra as input. This in meant to encourage the maintainers of those tools (I think mainly @timosachsenberg ) to change their SpectrumGenerator to this new one and switch to PeakSpectra instead of RichPeakSpectra completely.

timosachsenberg commented:
Having two implementations in the transition phase sounds good to me. We just need to make sure to rename and remove the old one when we are done. But in this case I see no problem. I am happy to adapt SimpleSearchEngine, RNPxlSearch, and potentially others.

hroest commented:
in that case I would suggest to create a new namespace `OpenMS::Deprecated` and put the old TSG in there. That makes it easy for the transition period and we can make sure we catch all instances.

timosachsenberg commented:
can you add me as a collaborator in github?

timosachsenberg commented:
Hi Eugen I also did a small benchmarking on our nodes:

Debug mode:
1e6 spectra in 18.535 s new TSG (53952 spectra / s)
1e6 spectra in 20.660 s old TSG (48402.7 spectra /s)

Release:
1e6 spectra in 3.241 sec new TSG (308547 spectra / s)
1e6 spectra in 3.790 sec old TSG (263852 spectra / s)

So it is pretty cool that the speedup translates to the release mode.

timosachsenberg commented:
I replaced all occurences of the old TSG with the new one. Then I promoted the new TSG. It is now the default one and the old was removed.

Before I did so, I validated that the new implementation yielded the same identifications using SimpleSearchEngine and a 5000 spectra dataset.

```
SimpleSearchEngine took 01:29 m (wall), 01:29 m (CPU), 0.00 s (system), 01:29 m (user).
<While loading 'QExactive1_MS2_2+_#5000.mzML': Name of CV term not correct: 'MS:1000422 - high-energy collision-induced dissociation' should be 'beam-type collision-induced dissociation'> occurred 5000 times
Mapping 4881 peptides to 173450 proteins.

Aho-Corasick done:
  found 18187 hits for 4881 of 4881 peptides (time: 15.424498 s (wall), 15.370006 s (CPU)).
Peptide hits passing enzyme filter: 18187
     ... rejected by enzyme filter: 0
-----------------------------------
Peptides statistics

  target/decoy:
    match to target DB only: 3854
    match to decoy DB only : 1014
    match to both          : 13

  mapping to proteins:
    no match (to 0 protein)         : 0
    unique match (to 1 protein)     : 1205
    non-unique match (to >1 protein): 3676
-----------------------------------
Protein statistics

  new proteins: 10925
  orphaned proteins: 0 (all removed)
PeptideIndexer took 19.02 s (wall), 18.66 s (CPU), 0.00 s (system), 18.66 s (user).
Omitted writing of 1014 peptide identifications due to empty hits.
FalseDiscoveryRate took 1.62 s (wall), 1.53 s (CPU), 0.00 s (system), 1.53 s (user).
Filtering by peptide score...
Removing unreferenced protein hits...
Before filtering:
1 protein identification(s) with 10925 protein hit(s),
3867 peptide identification(s) with 3867 peptides hit(s).
After filtering:
1 protein identification(s) with 4888 protein hit(s),
2476 peptide identification(s) with 2476 peptides hit(s).
IDFilter took 0.88 s (wall), 0.83 s (CPU), 0.00 s (system), 0.83 s (user).
```
These are identical to the ones produced before.

timosachsenberg commented:
please wait with merge... think I need to add a fix

timosachsenberg commented:
And if we compare the run times:
old:
SimpleSearchEngine took 02:14 m (wall), 02:12 m (CPU), 0.00 s (system), 02:12 m (user).
new:
SimpleSearchEngine took 01:29 m (wall), 01:29 m (CPU), 0.00 s (system), 01:29 m (user).

Which is a particularly nice speed-up. So, thanks a lot for your contribution. I need to fix a test first but I think we are close to a merge.

timosachsenberg commented:
I realized that there is some more work to be done.
Basically every template function that takes either a PeakSpectrum or RichPeakSpectrum (e.g., most of our filter) needs to be adapted. The reason is that previously push_back or copy was used to filter peaks (and DataArrays were not considered at all). Now we really need to consider DataArrays and use the select function of MSSpectrum

timosachsenberg commented:
I still get a small discrepancy in localization scores in the RNPxlSearch test. Need still to figure out if this is a bug or not...

hroest commented on src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:
```diff
@@ -743,6 +743,18 @@ namespace OpenMS
+    if (add_metainfo)
+    {
+      if (spectrum.getIntegerDataArrays().size() == 0)
+      {
```
what happens if they are not zero? Why not simply add one
```
spectrum.getIntegerDataArrays().resize(spectrum.getIntegerDataArrays().size() +1);
```

timosachsenberg replied:
There are other cases to consider: For example, what happens if the last data array has the same name? Add another one or put values in it?
Right now, I decided for the simplest and less ambiguous solution: just add array if none is in there.
But I can see the point that we might need to come up with a policy or "good practice" guide. Question is how we should proceed here?

hroest replied:
I think it would be somewhat "surprising" if you call a function that is supposed to add this kind of meta-data and then simply decides to not actually add any data

hroest commented on src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:
```diff
@@ -961,24 +972,22 @@ namespace OpenMS
           {
-            p_.setMetaValue("IonName", ion_name);
+            spectrum.getStringDataArrays()[0].push_back(ion_name);
+            spectrum.getIntegerDataArrays()[0].push_back(charge);
```
you could change this to 
```
+            spectrum.getStringDataArrays()[ion_name_array].push_back(ion_name);
+            spectrum.getIntegerDataArrays()[ion_name_array].push_back(charge);
```

(see above) instead of assuming that its always the first array

hroest commented on src/openms/source/CHEMISTRY/SvmTheoreticalSpectrumGenerator.cpp:
```diff
@@ -961,24 +972,22 @@ namespace OpenMS
         {
-          p_.setMetaValue("IonName", ion_name);
+          spectrum.getStringDataArrays()[0].push_back(ion_name);
+          spectrum.getIntegerDataArrays()[0].push_back(charge);
```
same here...

hroest commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -174,10 +182,11 @@ namespace OpenMS
-      spec.push_back(p);
+        String ion_name("iH");
+        spectrum.getStringDataArrays()[0].push_back(ion_name);
+        spectrum.getIntegerDataArrays()[0].push_back(1);
```
same here

hroest commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -187,10 +196,11 @@ namespace OpenMS
-        p.setMetaValue("IonName", name);
+        String ion_name("iF");
+        spectrum.getStringDataArrays()[0].push_back(ion_name);
+        spectrum.getIntegerDataArrays()[0].push_back(1);
```
same here

mwalzer commented:
So, what's the status of this?

timosachsenberg commented:
this PR is basically done but some discussion on policy how dataarrays are handled if others are present.
One solution is to always add the new one.
Another one is to allow only spectra withou dataarrays as input for certain algorithms.
One could modify a datarray if one with the same name is already present or throw an error. Last two solutions are very slow and probably not suitable for theo. spectra generation.
comments welcome

enetz commented:
So is there anything else we can do here?
I named the created dataarrays "Charges" and "IonNames", so that downstream tools can look for the right arrays. But I do not think it makes sense to check for these names in the generators themselves. They are meant to generate a new spectrum and should not be used to modify a spectrum, that already contains some other dataarrays. As long as the functions in the generators, like getSpectrum and addPeaks, are compatible with each other, we should not overdesign this.

All those "if (spectrum.getIntegerDataArrays().size() == 0)" are meant to check, whether this is a completely new spectrum or if some peaks with metainfo were added by getSpectrum or addPeaks already.
We assume, that these spectra were only ever touched by one of these functions and were not modified elsewhere, so if there are dataarrays we can be sure they are the ones we need and we can add the new info to them.

Actually, if we make addPeaks(...) protected and only allow getSpectrum(...) to be used by other tools, we could probably simplify and speed up this code a little more, since we could handle most of this in getSpectrum once instead of checking everything each time we call addPeaks.
But this depends on how useful and necessary the addPeaks(...) function is outside of the generators themselves.

hroest commented:
I have nothing against well-documented and tested assumptions. However, it seems that addPeaks is used a few times in the code base:
```
$ grep -R addPeaks src/openms | cut -f 1 -d ":" | uniq -c
      4 src/openms/source/ANALYSIS/OPENSWATH/DIAHelper.cpp
      2 src/openms/source/ANALYSIS/ID/AScore.cpp
      7 src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp
      1 src/openms/include/OpenMS/CHEMISTRY/TheoreticalSpectrumGenerator.h
```

So, my preferred solution is to try and make it as fast as possible, adding appropriate PRECONDITION  to check whether the caller did the correct thing and then go forward.This is explicit, clear and provides expectations that are actually checked at runtime (in debug mode at least).

Maybe a simple solution: the addPeaks function assumes that `add_metainfo_ == true` implies that there are two data arrays and uses a PRECONDITION macro to check for this. If we want to we could also pass the actual locations of the data arrays with two `Size` variables (default value = 1), maybe that is overkill? Otherwise its simply the job of the caller to provide a String and and Integer data array of exactly size 1.

timosachsenberg commented:
I think the only reason to use addPeaks is to be able to add a range of charges.
So I guess we could remove or hide them from the user and just expose a generate spectrum function that takes a charge range as an argument. Note that the current addPeaks is highly inefficient as it sorts all arrays every time.

timosachsenberg commented:
@enetz can you check if we can get rid of the addPeaks function without any disadvantage?

enetz commented:
It looks to me, that addPeaks and addPrecursorPeaks is mainly used as a shortcut for things, that can be achieved by using getSpectrum after changing the parameters of the TSG, like adding only y-ions. All we have to do is add a few lines to set the correct parameters and call getSpectrum at each place addPeaks is used right now. Although I am not sure how that would affect the performance.
The only exception are a few places, where addPeaks is used because it allows to set one specific charge instead of a maximal charge. For this we could add a parameter for a minimal charge to getSPectrum, which would enable specific ranges or specific values.

timosachsenberg commented:
Would be great if you could do this. I can try to assist but this will depend on my connection...

enetz commented:
@timosachsenberg I noticed, that the getSpectrum function only adds precursor peaks with one charge, the maximal charge. I guess this makes sense, since you would only expect the precursor mass to be present with the precursor charge in the MS2. Is that the correct assumption behind this behavior, or should the mass actually be added with all smaller charges as well?
At least in one case, in RNPxlSearch, you explicitly add precursor peaks for all charges smaller or equal to precursor charge. The problem now is to allow for both behaviors using the getSpectrum function. There are several ways to do this:
1. add a parameter for "add_all_precursor_charges"
2. we could remove the "spectrum.clear(true)" line at the beginning of getSpectrum and basically allow it to be used like addPeaks in a cumulative manner. I do not see it being used to overwrite a non-empty spectrum anywhere, so this should not affect any existing code, and even then you can just call "spectrum.clear(true)" before calling getSpectrum in those cases.

timosachsenberg commented:
I think adding the precursor peak is only necessary for the highest charge.
RNPxlSearch can also switch to the new behavior.

enetz commented:
@timosachsenberg are you sure about that? there is even a comment explicitly explaining what is done, so it seems it was done deliberately:
https://github.com/enetz/OpenMS/blob/feature/TSG_without_RichPeak/src/utils/RNPxlSearch.cpp#L995

I also found another place where something similar is done in TOPPView here:
https://github.com/enetz/OpenMS/blob/feature/TSG_without_RichPeak/src/openms_gui/source/VISUAL/TOPPViewIdentificationViewBehavior.cpp#L434

If this is not actually necessary, then I will make the change.

timosachsenberg commented:
I will check back with Aleksandar. Maybe this is specific to cross-link MS...

timosachsenberg commented:
I guess in this case it might make sense as this is the precursor after loss of the RNA. So actually don't know if also some charge was lost.

timosachsenberg commented:
Yeah, so the option "add_all_precursor_charges" would make sense. In TOPPView, I think we can deal with the default case of only adding the highest precursor.

enetz commented:
I added a new parameter "add_all_precursor_charges" to the TSG.
getSpectrum has two charge parameters, min_charge and max_charge.
Can be used to define a range or one specific charge if they are equal.
addPeaks, addPrecursorpeaks and addAbundantImmoniumIons are now protected,
but almost everything can be done with proper Param handling and the new charge range with getSpectrum. 
E.g. if you only want precursor ions, just set all 6 "add_x_ions" Parameters to false and "add_precursor_ions" to true.
I also removed the "spectrum.clear(true)" line from getSpectrum, because in all cases where this is necessary, the coders already did that before calling getSpectrum anyway, and this way it can also be used cumulatively like addPeaks before.

Do not merge yet, I still have to change how addPeaks and getSpectrum work together to hopefully make it simpler and more efficient.

timosachsenberg commented on src/openms/source/CHEMISTRY/TheoreticalSpectrumGenerator.cpp:
```diff
@@ -70,6 +70,9 @@ namespace OpenMS
     defaults_.setValidStrings("add_precursor_peaks", ListUtils::create<String>("true,false"));
 
+    defaults_.setValue("add_all_precursor_charges", "false", "Adds precursor peaks with all charges in the given range");
+    defaults_.setValidStrings("add_precursor_peaks", ListUtils::create<String>("true,false"));
```
should be setValidStrings("add_all_precursor_charges

timosachsenberg commented on src/tests/class_tests/openms/source/TheoreticalSpectrumGenerator_test.cpp:
```diff
@@ -315,6 +240,72 @@ START_SECTION(void getSpectrum(PeakSpectrum& spec, const AASequence& peptide, In
+  TEST_EQUAL(charge_counts[0], 0)
+  TEST_EQUAL(charge_counts[1], 0)
+  TEST_EQUAL(charge_counts[2], 30)
+
```
maybe add the abundant immonium ion test cases here

timosachsenberg commented on src/utils/RNPxlSearch.cpp:
```diff
@@ -985,18 +985,16 @@ class RNPxlSearch :
-          }
-
-          total_loss_spectrum.sortByPosition();
+          Param params;
```
might be quite costly to set this in the loop. Maybe create two spectrum generators outside the loop?

enetz commented:
I moved the checks for existing DataArrays into getSpectrum
and removed sorting of the spectrum from addPeaks and the other protected functions.
All this is now done once in the getSpectrum function, instead of being repeated each time a group of peaks has been added by addPeaks or a similar function.

The code is simpler and runtime was slightly reduced from 84.40 sec to 80.85 sec (Release) or from 343.34 sec to 269.88 (Debug) for 1e4 very complex spectra containing all possible fragment types and annotations.
For only b/y-ions without metainfo, 1e6 spectra (Release):
from 9.88 sec to 8.42 sec.

enetz commented:
This PR is ready for merging now, if there are no other issues.

hroest commented on src/openms/include/OpenMS/ANALYSIS/MRM/MRMFragmentSelection.h:
```diff
@@ -79,12 +79,12 @@ namespace OpenMS
 
     /// selects accordingly to the parameters the best peaks of spec and writes them into selected_peaks
-    void selectFragments(std::vector<RichPeak1D> & selected_peaks, const RichPeakSpectrum & spec);
+    void selectFragments(std::vector<Peak1D> & selected_peaks, const PeakSpectrum & spec);
```
please adjust PyOpenMS pxd

hroest commented on src/openms/include/OpenMS/ANALYSIS/RNPXL/HyperScore.h:
```diff
@@ -60,7 +60,9 @@ struct OPENMS_DLLAPI HyperScore
-  static double compute(double fragment_mass_tolerance, bool fragment_mass_tolerance_unit_ppm, const PeakSpectrum& exp_spectrum, const RichPeakSpectrum& theo_spectrum);
+//  static double compute(double fragment_mass_tolerance, bool fragment_mass_tolerance_unit_ppm, const PeakSpectrum& exp_spectrum, const RichPeakSpectrum& theo_spectrum);
+
+  static double compute(double fragment_mass_tolerance, bool fragment_mass_tolerance_unit_ppm, const PeakSpectrum& exp_spectrum, const PeakSpectrum& theo_spectrum);
```
please adjust PyOpenMS pxd

hroest commented on src/openms/include/OpenMS/CHEMISTRY/SvmTheoreticalSpectrumGenerator.h:
```diff
@@ -213,7 +213,7 @@ namespace OpenMS
 
     /// Generate the MS/MS according to the given probabilistic model
-    void simulate(RichPeakSpectrum & spectrum, const AASequence & peptide, boost::random::mt19937_64& rng, Size precursor_charge);
+    void simulate(PeakSpectrum & spectrum, const AASequence & peptide, boost::random::mt19937_64& rng, Size precursor_charge);
```
please adjust PyOpenMS pxd

hroest commented on src/openms/include/OpenMS/CHEMISTRY/SvmTheoreticalSpectrumGeneratorSet.h:
```diff
@@ -81,7 +81,7 @@ namespace OpenMS
 
     /// Generate the MS/MS according to the model for the given precursor_charge
-    void simulate(RichPeakSpectrum & spectrum, const AASequence & peptide, boost::random::mt19937_64& rng, Size precursor_charge);
+    void simulate(PeakSpectrum & spectrum, const AASequence & peptide, boost::random::mt19937_64& rng, Size precursor_charge);
```
please adjust PyOpenMS pxd

hroest commented on src/openms/include/OpenMS/KERNEL/MSSpectrum.h:
```diff
@@ -728,6 +731,12 @@ namespace OpenMS
+        {
+          throw Exception::Precondition(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "FloatDataArray[" + String(i) + "] size (" + 
+            String(float_data_arrays_[i].size()) + ") does not match spectrum size (" + String(peaks_old) + ")");
+        }
```
there are cases where the float/double arrays may not be of the same lengths as the rest of the spectrum -- maybe we just say then you cannot use `MSSpectrum::select` ?

timosachsenberg replied:
is this allowed by the mzML standard (sorry can't check for myself)?

hroest replied:
yes it is, we actually just recently had a discussion about it and added cv terms that are specifically designed for this:

https://sourceforge.net/p/psidev/mailman/psidev-ms-vocab/thread/a797e011d942e997f15d94c5edf55bcc%40mail.gmail.com/#msg35673878

so there are at least three terms that have "subsampled" data in them

hroest commented on src/tests/class_tests/openms/source/HyperScore_test.cpp:
```diff
@@ -79,24 +79,22 @@ START_SECTION((static double compute(double fragment_mass_tolerance, bool fragme
-  TEST_REAL_SIMILAR(HyperScore::compute(0.1, false, exp_spectrum, theo_spectrum), 7.39693);
-  TEST_REAL_SIMILAR(HyperScore::compute(10, true, exp_spectrum, theo_spectrum), 7.39693);
+  //TEST_REAL_SIMILAR(HyperScore::compute(0.1, false, exp_spectrum, theo_spectrum), 3.609438);
+  //TEST_REAL_SIMILAR(HyperScore::compute(10, true, exp_spectrum, theo_spectrum), 3.609438);
```
why commented out?

hroest commented:
generally looks good to me, please change the pxd files for the the pyOpenMS bindings accordingly when you changed the public interface

enetz commented:
I adapted the remaining pyOpenMS bindings.
Merged develop to solve issues in compiling pyOpenMS.

The issue with the HyperScore test is not yet resolved.
@timosachsenberg what should we do with the HyperScore tests, or rather how did they work before?
Did the HyperScore always just return 0, if there is no annotation?
The annotations are used for the y and bFactors, but the dot_product is computed without them,
should we just use only the dot_product, if there is no annotation available?
I think that is what I did anyway, since my annotations from XL-MS were not counted.

timosachsenberg commented:
hmm I think it is better to fail if no annotations are present and provide a class that calculates the dot product score instead.

timosachsenberg commented:
returning 0 would also be an alternative

timosachsenberg commented:
ready for merge or more to come?

enetz commented:
@timosachsenberg  @hroest 
The only issue left should be allowing DataArrays, that have a different length, than the spectrum.
I am not sure how to handle this. We could just exclude these cases from all sorting and filtering algorithms etc.
But what if such a DataArray, that has no direct relation to the number or order of peaks, has the same number of peaks by accident? Should it then be sorted anyway? I think to allow that we would need an additional flag for each Array or something like that.
Since we do not use such algorithms yet, we could merge this for now and open another issue to discuss that specifically.

timosachsenberg commented:
+1

hroest commented:
@enetz @timosachsenberg I agree, throwing an exception is an acceptable solution as it is not clear how the software should deal with that case
